lib/spack/spack/relocate.py

```diff
@@ -415,6 +426,17 @@ def make_binary_placeholder(cur_path_names, allow_root):
         tty.die("Placeholder not implemented for %s" % platform.system())
 
 
+def relocate_link(path_names, old_dir, new_dir):
+    """
+    Replace old path with new path in link sources
+    """
+    for path_name in path_names:
+        old_src = os.path.realpath(path_name)
```

Comment:
Note that for links that are already relative, depending on where you are when you do this, the result can be incorrect, in the sense that it constructs the path by appending the link target to your CWD (that's primarily a failing of Python's link resolution logic). I recommend only performing this sort of transformation on links that refer to absolute paths (unless I overlooked a pre-filtering step which happens before this function).

For more info, see `llnl.util.filesystem.resolve_link_target_relative_to_the_link`

There are a couple other references to `os.path.realpath` here and I have similar concerns for each one.